File: server/controllers/authController.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const Student = require('../models/Student');
const Admin = require('../models/Admin');

exports.register = async (req, res) => {
  try {
    const { name, email, password, skills, interests, role } = req.body;

    if (role === 'admin') {
      const existingAdmin = await Admin.findOne({ email });
      if (existingAdmin) {
        return res.status(400).json({ message: 'Admin already exists' });
      }

      const hashedPassword = await bcrypt.hash(password, 10);
      const admin = new Admin({ email, password: hashedPassword });
      await admin.save();
      return res.status(201).json({ message: 'Admin registration successful' });
    }

    const existingStudent = await Student.findOne({ email });
    if (existingStudent) {
      return res.status(400).json({ message: 'Student already exists' });
    }

    const hashedPassword = await bcrypt.hash(password, 10);

    const student = new Student({
      name,
      email,
      password: hashedPassword,
      skills: skills || [],
      interests: interests || []
    });

    await student.save();
    res.status(201).json({ message: 'Registration successful' });
  } catch (error) {
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

exports.login = async (req, res) => {
  try {
    const { email, password } = req.body;

    const admin = await Admin.findOne({ email });
    if (admin) {
      const isMatch = await bcrypt.compare(password, admin.password);
      if (!isMatch) {
        return res.status(400).json({ message: 'Invalid credentials' });
      }

      const token = jwt.sign(
        { id: admin._id, email: admin.email, role: 'admin' },
        process.env.JWT_SECRET,
        { expiresIn: '7d' }
      );

      return res.json({
        token,
        role: 'admin',
        admin: {
          id: admin._id,
          name: admin.name,
          email: admin.email,
          profilePic: admin.profilePic
        }
      });
    }

    const student = await Student.findOne({ email });
    if (!student) {
      return res.status(400).json({ message: 'Invalid credentials' });
    }

    const isMatch = await bcrypt.compare(password, student.password);
    if (!isMatch) {
      return res.status(400).json({ message: 'Invalid credentials' });
    }

    const token = jwt.sign(
      { id: student._id, email: student.email, role: 'student' },
      process.env.JWT_SECRET,
      { expiresIn: '7d' }
    );

    res.json({
      token,
      role: 'student',
      student: {
        id: student._id,
        name: student.name,
        email: student.email,
        profilePic: student.profilePic,
        skills: student.skills,
        interests: student.interests
      }
    });
  } catch (error) {
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

exports.getProfile = async (req, res) => {
  try {
    const Model = req.user.role === 'admin' ? Admin : Student;
    const profile = await Model.findById(req.user.id).select('-password');
    res.json(profile);
  } catch (error) {
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

exports.updateProfile = async (req, res) => {
  try {
    const { name, skills, interests, profilePic } = req.body;
    const Model = req.user.role === 'admin' ? Admin : Student;
    const profile = await Model.findByIdAndUpdate(
      req.user.id,
      { name, skills, interests, profilePic },
      { new: true }
    ).select('-password');

    res.json(profile);
  } catch (error) {
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

File: server_java/src/main/java/com/aspirepath/controller/AuthController.java
package com.aspirepath.controller;

import com.aspirepath.dto.AuthResponse;
import com.aspirepath.model.Admin;
import com.aspirepath.model.Student;
import com.aspirepath.repository.AdminRepository;
import com.aspirepath.repository.StudentRepository;
import com.aspirepath.security.CustomUserDetails;
import com.aspirepath.service.AuthService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    private final AuthService authService;
    private final StudentRepository studentRepository;
    private final AdminRepository adminRepository;

    public AuthController(AuthService authService, StudentRepository studentRepository, AdminRepository adminRepository) {
        this.authService = authService;
        this.studentRepository = studentRepository;
        this.adminRepository = adminRepository;
    }

    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody Map<String, Object> body) {
        try {
            @SuppressWarnings("unchecked")
            java.util.List<String> skills = (java.util.List<String>) body.get("skills");
            @SuppressWarnings("unchecked")
            java.util.List<String> interests = (java.util.List<String>) body.get("interests");
            
            authService.register(
                (String) body.get("name"), 
                (String) body.get("email"), 
                (String) body.get("password"), 
                (String) body.get("role"),
                skills,
                interests
            );
            return ResponseEntity.status(201).body(Map.of("message", "Registration successful"));
        } catch (Exception e) {
            return ResponseEntity.status(400).body(Map.of("message", e.getMessage()));
        }
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody Map<String, String> body) {
        try {
            AuthResponse response = authService.login(body.get("email"), body.get("password"));
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            return ResponseEntity.status(400).body(Map.of("message", e.getMessage()));
        }
    }

    @GetMapping("/profile")
    public ResponseEntity<?> getProfile(@AuthenticationPrincipal CustomUserDetails userDetails) {
        if ("admin".equals(userDetails.getRole())) {
            return ResponseEntity.ok(adminRepository.findById(userDetails.getId()).get());
        } else {
            return ResponseEntity.ok(studentRepository.findById(userDetails.getId()).get());
        }
    }

    @PutMapping("/profile")
    public ResponseEntity<?> updateProfile(@AuthenticationPrincipal CustomUserDetails userDetails, @RequestBody Map<String, Object> body) {
        if ("admin".equals(userDetails.getRole())) {
            Admin admin = adminRepository.findById(userDetails.getId()).get();
            if (body.containsKey("name")) admin.setName((String) body.get("name"));
            if (body.containsKey("profilePic")) admin.setProfilePic((String) body.get("profilePic"));
            return ResponseEntity.ok(adminRepository.save(admin));
        } else {
            Student student = studentRepository.findById(userDetails.getId()).get();
            if (body.containsKey("name")) student.setName((String) body.get("name"));
            if (body.containsKey("profilePic")) student.setProfilePic((String) body.get("profilePic"));
            if (body.containsKey("skills")) {
                @SuppressWarnings("unchecked")
                java.util.List<String> skills = (java.util.List<String>) body.get("skills");
                student.setSkills(skills);
            }
            if (body.containsKey("interests")) {
                @SuppressWarnings("unchecked")
                java.util.List<String> interests = (java.util.List<String>) body.get("interests");
                student.setInterests(interests);
            }
            return ResponseEntity.ok(studentRepository.save(student));
        }
    }
}
